### ISSUE TITLE
Stream FPS suffers immensely when streaming intensive games

### ISSUE THREAD

**Issue by ockerman0** [maltejur/discord-screenaudio#50]
I took interest in this as I like to stream games to my friend group sometimes but the official client still doesn't support audio. For 3/4 of the games I tested, the stream's FPS was extremely low when wandering in areas that otherwise show up just fine on the official client. As expected, choosing lower streaming resolutions helps alleviate this, but it doesn't go away completely.

**Comment by Maciejka1** [comment]
I think it's because lack of hardware acceleration on discord web version and desktop version on linux.

**Comment by SvdB-nonp** [comment]
Just wanted to raise this as well. CPU usage increases from 30 to around 75% when streaming, so GPU is indeed unused.

Is there a way to enable hardware acceleration? Since it is electron/chrome based, are there flags one can enable for VA-API or whatever it is called?

**Comment by maltejur** [comment]
Judging from https://wiki.qt.io/QtWebEngine/VideoAcceleration I don't think this can be fixed easily, but I could be wrong

**Comment by Kiamrt** [comment]
i am not very good at any of this but from the documentation in the QtWebEngine wiki maltejur sent there is this https://codereview.qt-project.org/c/qt/qtwebengine-chromium/+/209204/

**Comment by Sirmentio** [comment]
I feel as though this issue should get more attention, this can be a real deal breaker even if there is working audio. For viewers, having the stream at an uncomfortable stutter dependent on hardware can be a real drag for watch-ability.

**Comment by maltejur** [comment]
Hm judging from the qt bug report there is not much I can do about that. But I have found one promising solution/hack in [a issue from qutebrowser](https://github.com/qutebrowser/qutebrowser/issues/2671).

I have put my code on the [testing_hardware_accel](https://github.com/maltejur/discord-screenaudio/tree/testing_hardware_accel) branch. Can somebody who experiences lags try out that code and see if it gets any better?

**Comment by ockerman0** [comment]
> 

Hi, I've just tried the client, built from that branch. There seems to be no difference in performance between it and the main branch, the performance drop still persists.

**Comment by Sirmentio** [comment]
I can't quite get it to work myself (For me, it just goes to a blank page, though I might have not compiled it right.). I'm using Wayland

**Comment by Sirmentio** [comment]
I have tested it myself, and from people viewing my stream, it's still very choppy.

**Comment by maltejur** [comment]
Hm ok then there is nothing I can really do.

**Comment by Kiamrt** [comment]
uhhh i checked out the flags that you added, but some are missing
for example those are the ones i use with my chrome/chromium
--enable-accelerated-2d-canvas
--enable-accelerated-video-decode
--enable-accelerated-mjpeg-decode
--enable-accelerated-video-encode
i'm pretty sure those are needed for it to somehow work correct me if i'm wrong

**Comment by maltejur** [comment]
You can test if you can get hardware acceleration running by setting `QTWEBENGINE_CHROMIUM_FLAGS` to those chromium flags.

**Comment by rowbawts** [comment]
The interesting part is that even with what I consider high end specs my stream also has complaints of lag from viewers even just in the browser playing video at times let alone gaming.

```
CPU: AMD Ryzen 5800X
GPU: AMD 6700XT
Using Flatpak discord-screenaudio
```

**Comment by maltejur** [comment]
@rowbawts are you sure this is not caused by your internet connection? Do you have a high CPU usage? If I stream 1080p on the official discord client my stream is also quite laggy.

**Comment by rowbawts** [comment]
> @rowbawts are you sure this is not caused by your internet connection? Do you have a high CPU usage? If I stream 1080p on the official discord client my stream is also quite laggy.

If I use OBS I have no issues.

![image](https://user-images.githubusercontent.com/13012832/198139829-daf792be-5aed-4c24-9522-1d07f7839a2a.png)

**Comment by maltejur** [comment]
@rowbawts are you streaming on the same resolution and FPS in the official client and here?

But if this is indeed not the issue I see that something should be done here, I just don't know what. The official Qt wiki states that video acceleration is not supported and testing out some command line flags has not worked for me either.

**Comment by rowbawts** [comment]
@maltejur Yeah everytime I do 1080p 60fps but even when trying other lower settings the same issue occurs. But I completely understand if you end up figuring out a way to enable hw accel it should definitely improve.

**Comment by Kiamrt** [comment]
![image](https://user-images.githubusercontent.com/114337606/200122095-0b7040b0-6184-4590-a5cb-7507780d7223.png)
i got hardware acceleration to work, sadly for decoding only, those are the flags i used, also bear in mind i'm using the va-api feature

curious how many of the flags are getting ignored, also, if i change num-raster-threads it doesn't get any higher

qputenv("QTWEBENGINE_CHROMIUM_FLAGS",
          "--ignore-gpu-blacklist --enable-gpu-rasterization "
          "--enable-accelerated-video-decode --enable-accelerated-video-encode "
          "--enable-native-gpu-memory-buffers --num-raster-threads=4 "
          "--enable-features=WebRTCPipeWireCapturer,VaapiVideoEncoder,VaapiVideoDecoder "
          "--disable-features=UseOzonePlatform,UseChromeOSDirectVideoDecoder,UseChromeOSDirectVideoEncoder " +
              qgetenv("QTWEBENGINE_CHROMIUM_FLAGS"));

I tested this by watching a video with lowered cpu clock speeds so usage would be higher, and monitoring my gpu through green with envy, sadly as i said this only works for hardware decoding, it seems

**Comment by AsakaJX** [comment]
I can't figure how could I set flags.

**Comment by IceDBorn** [comment]
Not a solution, but firefox works way better for me with intensive games. Inspired by this project, I created this [extension](https://github.com/IceDBorn/pipewire-screenaudio) for firefox. It is not packaged for any distro, I'm open to contributions though!

**Comment by TibixDev** [comment]
Still no news on working GPU hardware acceleration?

**Comment by aaomidi** [comment]
Two things I'm curious about. (Wayland Specific)

1. OBS somehow is able to capture the screen with no impact to the game, and at a high refresh rate. How?
2. Maybe discord-screenaudio can create a local rtmp server, and use the contents from there to stream to discord? That way we can use OBS to actually capture the frames (?)

**Comment by chrisbrasington** [comment]
Things got a little weird lately.

flatpak vanilla discord `com.discordapp.Discord` started is working for screensharing video from individual games/windows quite well. Hardware Acceleration and video codec OPENH264 are enabled - which results in a smooth video stream. Problem is no audio.

flatpak `discord-screenaudio` while the audio is working, does not show hardware acceleration. Problem is unwatchable choppy video. 

Too bad I can't get both, smooth video and audio. I don't know enough about what's going on, but for video I need hardware acceleration / H264 to be watchable. 

I've tried this in both X11/wayland scenarios. GPU is `AMD 5700 XT`

![image](https://github.com/maltejur/discord-screenaudio/assets/1186896/dfdc2cde-d1d9-47a4-9bfd-037b2bf94662)
